README.md
# Protocol::ACME

See client.pl for an example of using the library.

# Status

This module is very rough right now.  The goal is to get feedback on key items such as:

* Naming
* Challenge handling
* Exception handling 

# Usage

## Requisite file creation

In order to use the library and the Let's Encrypt service in general you will need to generate a couple of files.  These are:

* A private account key 
* A private cetificate key
* A certificate signing request ( CSR )

There are numerous ways to go about this.  Below are some command line recipies using OpenSSL.

1) Generate a new private key for the Let's Encrypt account: 
 
 `$ openssl genrsa -out account_key.pem 2048`
 
2) Generate a new private key for the certificate:
 
 `$ openssl genrsa -out cert_key.pem 2048`

3a) Generate a certificate signing request (CSR).  For a single domain cert:

`$ openssl req -new -sha256 -key cert_key.pem -outform der -subj "/CN=cloud.example.com" > csr.der`

 3b) Generating a CSR for a SAN cert ( multiple domains ) is a bit more work.  Grab a version of openssl.cnf and add the following:

```
   [SAN]
   subjectAltName=DNS:domain1.example.com,DNS:domain2.example.com
```

  and then generate with something like:

`$ openssl req -new -out test.csr -outform der -key cert_key.pem -config openssl.cnf -reqexts SAN -subj "/CN=domain.example.com" -sha256`

  This will create a cert with three domains.  domain.example.com will be in the subject and
  domain1.example.com and domain2.example.com will be in the SAN extension.

# API Usage

The goal of this module is to take away much of the need to understand the details of the Let's Encrypt ACME api.  That said, having a rough understanding of the flow will be usefil.  client.pl was writted and commented in a way to document the API and the manner in which to use the module.

## Challenges

The goal of the challenges code is to make it as flexible as possible by having the user provide a closure / function reference that is passed the the ACME module.  Each identifier will need its own challenge fullfillment code.

TODO: Make this multiple modules that take care of different use cases such as dropping files locally, copying via SSH, pusting to github, etc.  FLexibility is the goal.



Changes
Revision history for Protocol-ACME

1.01    Fri Oct 14
        - Congratulations. Promoting to Version 1.00

1.00    Mysteriously Skipped

0.16    Wed Aug 10
        - Catches json decode errors and bubbles them up
        - Added the implicit SSL dependencies of HTTP::Tiny

0.15    Mon Jul 6
        - Changed tests so setting NO_NETWORK environment variable
          will skip the network tests direct to Let's Encrypt
        - Adjusted MANIFEST so that client.pl and revoke.pl will not
          be installed (moved to examples directory)

0.14    Mon Jul 6
        - Typos in ACME.pm POD

0.13    Mon Jul 6
        - No changes of consequence

0.12    Fri Apr 29
        - Added cleanup routines to take care of removing the challenge files
          when complete.
        - Fixed a bug where the 'mailto' scheme was not added to the registraion
          calls.
        - Fixed bug with the order of execution of setgid and setuid
        - Refactored logging clde
        - Lowered dependency version 
        - Typos

0.11    Fri Jan 29
        - Version bump to deal with a missing version in a number of files

0.10    Thu Jan 28
        - Added a 'chain' call
        - Added the ability in set the mailto argument

0.09    Tue Jan 19
        - Made the log level configurable
        - Added more tests
        
0.08_01 Mon Jan 18
        - Added flexibility calls that take a PEM/DER string/filename
        - Fixed various minor bugs
        - Changes to the SimpleSSH challenge to create target directory as needed
        - Updated docs
        - deprecated load_key_from_disk
        - changed the meaning of the account_key parameter lassed to the constructor
        - removed the account_key_path paramater from the constructor

0.08    Wed Jan 12
        - Moved from LWP to HTTP::Tiny

0.07    Mon Jan 10
        - Moved pem2der, der2pem to external module

0.06    Sun Jan  9
        - Fixed REPOSITIRY section of POD

0.05    Sat Jan  8
        - More flexibility between using PEM and DER
        - Keys can be passed in via buffer
        
0.04    Mon Jan  4 16:45:02 PST 2016
        - Added ability to chose between using a local openssl binary
          for crypto and Crypt::OpenSSL::RSA
        - Changed from the deprecated Digest::SHA2 to Digest::SHA

0.03    Added Revoke support

0.02    Changes lost in the mists of time

0.01    Tue Dec  8 17:01:36 PST 2015
        Inital version

        



Makefile.PL
use 5.006;
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME             => 'Protocol::ACME',
    AUTHOR           => q{Stephen Ludin <sludin@ludin.org>},
    VERSION_FROM     => 'lib/Protocol/ACME.pm',
    ABSTRACT_FROM    => 'lib/Protocol/ACME.pm',
    LICENSE          => 'artistic_2',
    PL_FILES         => {},
    MIN_PERL_VERSION => 5.007003,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    TEST_REQUIRES => {
        'Test::More' => 0,
        'Test::Exception' => 0,
    },
    META_MERGE => {
        'meta-spec' => { version => 2 },
         resources => {
             repository => {
                 type => 'git',
                 url  => 'https://github.com/sludin/Protocol-ACME.git',
                 web  => 'https://github.com/sludin/Protocol-ACME',
             },
         },
    },
    PREREQ_PM => {
                  # here the prerequisites are aligned with Debian
                  # stable, if they exists, otherwise to stretch (and
                  # they need backporting: libcrypt-rsa-parse-perl and
                  # libcrypt-format-perl
          'Log::Any' => '0.15',
          'Log::Any::Adapter' => '0.11',
          'Crypt::Format' => '0.04',
          'Crypt::OpenSSL::RSA' => '0.28',
          'Crypt::RSA::Parse' => '0.02',
          'Crypt::OpenSSL::Bignum' => '0.04',
          'JSON' => '2.61',
          'Digest::SHA' => '5.88',
          'HTTP::Tiny' => '0.050',
          'Test::Exception' => '0.35',
          'IO::Socket::SSL' => '1.56',
          'Net::SSLeay' => '1.49',
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'Protocol-ACME-*' },
    x_contributors => [ 'Felipe Gasper <felipe@felipegasper.com>',
                        'Tobias Oetiker <tobi@oetiker.ch>' ],
);

sub MY::c_o {
  return <<'MAKE_FRAG';
readme:
	podselect lib/Protocol/ACME.pm > README.pod
MAKE_FRAG
}


